src/sinks/gcp_pubsub.rs

```diff
@@ -0,0 +1,427 @@
+use crate::{
+    buffers::Acker,
+    event::Event,
+    sinks::util::{
+        http::{HttpRetryLogic, HttpService},
+        tls::{TlsOptions, TlsSettings},
+        BatchConfig, Buffer, SinkExt, TowerRequestConfig,
+    },
+    topology::config::{DataType, SinkConfig, SinkDescription},
+};
+use bytes::{BufMut, BytesMut};
+use futures::{stream::iter_ok, Future, Sink, Stream};
+use goauth::{auth::JwtClaims, auth::Token, credentials::Credentials, error::GOErr, scopes::Scope};
+use http::{Method, Uri};
+use hyper::{
+    header::{HeaderValue, AUTHORIZATION},
+    Body, Client, Request,
+};
+use hyper_tls::HttpsConnector;
+use lazy_static::lazy_static;
+use serde::{Deserialize, Serialize};
+use smpl_jwt::Jwt;
+use snafu::{ResultExt, Snafu};
+use std::sync::{Arc, RwLock};
+use std::time::Duration;
+use tokio::timer::Interval;
+
+#[derive(Deserialize, Serialize, Debug, Clone, Default)]
+#[serde(deny_unknown_fields)]
+pub struct PubsubConfig {
+    pub project: String,
+    pub topic: String,
+    pub emulator_host: Option<String>,
+    pub api_key: Option<String>,
+    pub credentials_path: Option<String>,
+
+    #[serde(default, flatten)]
+    pub batch: BatchConfig,
+    #[serde(flatten)]
+    pub request: TowerRequestConfig,
+
+    pub tls: Option<TlsOptions>,
+}
+
+#[derive(Debug, Snafu)]
+enum BuildError {
+    #[snafu(display("GCP pubsub sink requires one of api_key or credentials_path to be defined"))]
+    MissingAuth,
+    #[snafu(display("Invalid GCP credentials"))]
+    InvalidCredentials { source: GOErr },
+    #[snafu(display("Invalid RSA key in GCP credentials"))]
+    InvalidRsaKey { source: GOErr },
+    #[snafu(display("Failed to get OAuth token"))]
+    GetTokenFailed { source: GOErr },
+}
+
+inventory::submit! {
+    SinkDescription::new::<PubsubConfig>("gcp_pubsub")
+}
+
+lazy_static! {
+    static ref REQUEST_DEFAULTS: TowerRequestConfig = TowerRequestConfig {
+        ..Default::default()
```

Comment:
If no fields are set, maybe this could be shortened to

```rust
static ref REQUEST_DEFAULTS: TowerRequestConfig = Default::default();
```

?

Reply:
Oh yeah, if it's just the defaults we can even drop `REQUEST_DEFAULTS` entirely and just use default in the `unwrap_with` call. The question is if the defaults are appropriate for pubsub.